extensions/amp-fx-collection/0.1/providers/fade-in.js

```diff
@@ -14,43 +14,16 @@
  * limitations under the License.
  */
 
-import {
-  PositionObserverFidelity,
-} from '../../../../src/service/position-observer/position-observer-worker';
-import {Services} from '../../../../src/services';
+import {FxElement, FxProvider} from './fx-provider';
 import {dev} from '../../../../src/log';
-import {getServiceForDoc} from '../../../../src/service';
-import {
-  installPositionObserverServiceForDoc,
-} from '../../../../src/service/position-observer/position-observer-impl';
 import {setStyles} from '../../../../src/style';
 
 /**
  * Provides a fade-in visual effect.
  *
  * @implements {../amp-fx-collection.FxProviderInterface}
  */
-export class FadeInProvider {
-
-  /**
-   * @param  {!../../../../src/service/ampdoc-impl.AmpDoc} ampdoc
-   */
-  constructor(ampdoc) {
-
-    /** @private @const {!../../../../src/service/ampdoc-impl.AmpDoc} */
-    this.ampdoc_ = ampdoc;
-
-    /** @private @const {!../../../../src/service/viewport/viewport-impl.Viewport} */
-    this.viewport_ = Services.viewportForDoc(ampdoc);
-
-    /** @private @const {!../../../../src/service/resources-impl.Resources} */
-    this.resources_ = Services.resourcesForDoc(ampdoc);
-
-    installPositionObserverServiceForDoc(ampdoc);
-
-    /** @private @const {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
-    this.positionObserver_ = getServiceForDoc(ampdoc, 'position-observer');
-  }
+export class FadeInProvider extends FxProvider {
```

Comment:
Looking better!

One nit, though. This is a classic example of composition over inheritance. If we keep from allocating new classes for each fx, then we're making a big leap in performance and maintanability.

How about passing the interpolation function to the FxProvider constructor? That way you also only have to write tests for `FxProvider` but not for subclasses (since they wouldn't exist).

If you need more granular control over the provider then consider a mixin approach, where each Fx would be a root class composed into `FxProvider`.

Reply:
I don't understand how passing the function into a constructor would happen in this case. The interpolation function (`update_()` in this case) needs access to variables stored on the classes themselves.

Reply:
1. update() can take in the animation configuration params
2. Instead of interpolation being a function, it could be a class.